exercises/resistor-color-trio/canonical-data.json

```diff
@@ -60,6 +60,78 @@
         "value": 470,
         "unit": "kiloohms"
       }
+    },
+    {
+      "uuid": "ee7a1b79-8848-4b1a-bf50-498c6fab7f27",
+      "description": "Brown and red and red",
+      "property": "label",
+      "input": {
+        "colors": ["brown", "red", "red"]
+      },
+      "expected": {
+        "value": 1200,
```

Comment:
You often say 1.2 k ohms instead of 1200 ohms. 
This may seem like a small thing. Although becomes pretty apparent when you come up to resistors with 5 bands.
Writing 11300 ohms just becomes very long.
I am not an expert on how it should be written but Wikipedia uses the style I mentioned.
I know some people who are better (than me) at electrical engineering to double-check this style.

Reply:
The goal of this exercise is not to produce common parlance. It is to convert strings to numbers and back to strings.

The units are all integers and is makes sense to remain consistent with the existing tests in this file, as well as with the existing tests in tracks which already use this data.

Reply:
For the first part these changes you are bringing up will brick more or less every tracks solutions.

Meaning it doesnt matter.

For the second part. A lot of people have solved this exercise by using:

If x > 1000
 X / 1000

This goes right in hand with my suggestion.

Your way would mean:

If x % 1000
   X / 1000

And that solution have I not seen in more or less any language I have mentored in ever nor as a community solution.

I am just saying that If this is suppose to be a "realstic" problem with a "realstic" output it should use float values and the way of converting a float to a string in a lot of languages arent different to int.
What is also important to mention is the fact that If there every comes a resistor color quadro. So would that one require the use of float becuase the forth band tells the accuracy.

Reply:
> For the first part these changes you are bringing up will brick more or less every tracks solutions.
> Meaning it doesnt matter.

I disagree. The problem spec says existing tests cannot be changed. These changes adds additional tests but do not change the behavior of existing tests. Track maintainers can opt to sync these additional tests or not.

> For the second part. A lot of people have solved this exercise by using:
> If x > 1000
> X / 1000

Can you quantify that assertion? Is that 0.5% of the solutions? 50%? I know that I would push back against seeing that in any code I mentor.

> Your way would mean:
> If x % 1000
> X / 1000
> And that solution have I not seen in more or less any language I have mentored in ever nor as a community solution.

In Python, I'd recommend `divmod()`. I've seen modulus used heavily in all the solutions I've read.

> I am just saying that If this is suppose to be a "realstic" problem with a "realstic" output it should use float values and the way of converting a float to a string in a lot of languages arent different to int.

Pros for this approach:
* These changes are consistent with the existing solutions.
* These changes are in line with existing changes with existing tracks.
* These changes are in line with what seems reasonable to me.

Cons for these changes:
* You think some people might not talk this way.

Is there any technical disadvantage or objection to these changes?

> What is also important to mention is the fact that If there ever comes a resistor color quadro. So would that one require the use of float becuase the forth band tells the accuracy.

This exercise is explicitly about three bands. A new exercise could handle the fourth band however it wants. That's not a technical or quantitative reasoning which has any relevancy to this change.

Reply:
Hmm. The verifier for this problem https://github.com/petertseng/exercism-problem-specifications/blob/verify/exercises/resistor-color-trio/verify.rb uses `>=` and `/`, because that's what made the most sense to me personally at the time, writing it. Previously the test cases did not distinguish between these following two rules:

1. If it is possible to express the number as a number of kiloohms greater than 1, express the number in kiloohms. Otherwise, express it in ohms.
2. Expess the number using the smallest SI prefix in {(none), kilo, mega, giga} that results in no fractional part.

So either was acceptable, and I happened to pick the first (which is not to say that the first is better or worse, just that it was a product of circumstance). To clarify, this test case's intention is to make explicit that only the second rule is acceptable, going forward? I will adjust accordingly if I can just get that confirmed, thank you.

Reply:
Hmm. The verifier for this problem https://github.com/petertseng/exercism-problem-specifications/blob/verify/exercises/resistor-color-trio/verify.rb uses `>=` and `/`, because that's what made the most sense to me personally at the time, writing it. Previously the test cases did not distinguish between these following two rules:

1. If it is possible to express the number as a number of kiloohms greater than or equal to 1, express the number in kiloohms. Otherwise, express it in ohms.
2. Expess the number using the largest SI prefix in {(none), kilo, mega, giga} that results in no fractional part (and therefore the smallest magnitude of value).
(a previous version of this message said "smallest", which didn't make sense. it should be "largest")

So either was acceptable, and I happened to pick the first (which is not to say that the first is better or worse, just that it was a product of circumstance). To clarify, this test case's intention is to make explicit that only the second rule is acceptable, going forward? I will adjust accordingly if I can just get that confirmed, thank you.

Reply:
@petertseng  Yes, please. That would be wonderful and consistent with the tracks were I've seen these extended tests.

Reply:
I **strongly** advise that if the second rule is the correct rule going forward that https://github.com/exercism/problem-specifications/blob/main/exercises/resistor-color-trio/description.md **must** be changed, because it currently says this:

> When we get more than a thousand ohms, we say "kiloohms". That's similar to saying "kilometer" for 1000 meters, and "kilograms" for 1000 grams.

This implies that you **should** use `>= 1000` to decide whether to use kiloohms. Therefore, if this test case is added, it will contradict the description unless the description changes, and we do not want to contradict ourselves. It can't say "more than a thousand ohms" anymore; it needs to mention "a whole number of kiloohms" in order to not violate the rule.

The description will of course also need to be updated to mention mega and giga.

Reply:
Thanks for noticing that, Peter! I updated the description. I'm hoping this doesn't make it overly verbose.

Reply:
I am leaving a replay to your earlier message Isaac.

> I disagree. The problem spec says existing tests cannot be changed. These changes adds additional tests but do not change the behavior of existing tests. Track maintainers can opt to sync these additional tests or not.

My way requires no current test to be changed.

> Can you quantify that assertion? Is that 0.5% of the solutions? 50%? I know that I would push back against seeing that in any code I mentor.

I will do a check if YOUR solution bricks it for people or for me in the most started solutions for: Javascript, Typescript, Python and Ruby.
Here is the result
Where my solution breaks: 25
Where Isaacs breaks: 36

Analsysis:

Why the numbers are so "low"? That is because a lot of exercises bricked for both of the new solutions. 
I also was fair in my choice of languages since how python and ruby deal with floats will brick my suggestion in all of those exercises more or less.
Even with that said your bricked 11 more solutions than my way.

> In Python, I'd recommend divmod(). I've seen modulus used heavily in all the solutions I've read.

With the first part, it sounds like you try to lecture me?
The exercise has only been out for a few days? None of the most starred solutions uses modulus. Therefore I am not sure what you are talking about. When I searched for the analysis I perhaps search 100+ solutions and I saw perhaps 2-4 use modulus.

> These changes are consistent with the existing solutions.

Also what are you basing this on???

> These changes are in line with existing changes with existing tracks.

I just told you my changes would brick fewer solutions. I did those tests for some of exercism that used languages aswell.

Pros for my approach:

- It is realistic in how you handle metric units.

- Breaks way fewer current solutions (not counting the addition of Giga and etc..)

- Would make it consistent with a further resistor color Quadro if it ever comes.

- Doesn't break any of the current problem spec

Cons for my approach:

- Would brick the solutions for the track awk


> This exercise is explicitly about three bands. A new exercise could handle the fourth band however it wants. That's not a technical or quantitative reasoning which has any relevancy to this change.

I think it does since you can use part of the resistor color dous solution for trio. This would make that impossible.

Reply:
@iHiD could I impose on you to weigh in, please?